libbpf-rs/src/map.rs

```diff
@@ -409,16 +426,81 @@ impl Map {
     pub fn keys(&self) -> MapKeyIter {
         MapKeyIter::new(self, self.key_size())
     }
+
+    /// Create the bpf map standalone.
+    pub fn create_map(
+        name: &str,
+        map_type: MapType,
+        key_size: u32,
+        value_size: u32,
+        max_entries: u32,
+        flags: MapCreateFlags,
+    ) -> Result<Map> {
+        let opts = libbpf_sys::bpf_map_create_opts {
+            sz: std::mem::size_of::<libbpf_sys::bpf_map_create_opts>() as libbpf_sys::size_t,
+            map_flags: flags.bits as i32,
+            btf_fd: 0,
+            btf_key_type_id: 0,
+            btf_value_type_id: 0,
+            btf_vmlinux_value_type_id: 0,
+            inner_map_fd: 0,
+            map_extra: 0,
+            numa_node: 0,
+            map_ifindex: 0,
+        };
+
+        let map_name_c = util::str_to_cstring(name)?;
+        let fd = unsafe {
+            libbpf_sys::bpf_map_create(
+                map_type.clone() as u32,
+                map_name_c.as_ptr(),
+                key_size,
+                value_size,
+                max_entries,
+                &opts,
+            )
+        };
+
+        if fd < 0 {
+            return Err(Error::System(fd));
+        }
+
+        Ok(Map {
+            fd,
+            name: name.to_string(),
+            ty: map_type as u32,
+            key_size,
+            value_size,
+            ptr: std::ptr::null_mut(),
+        })
+    }
 }
 
 #[rustfmt::skip]
 bitflags! {
     /// Flags to configure [`Map`] operations.
     pub struct MapFlags: u64 {
-	const ANY      = 0;
-	const NO_EXIST = 1;
-	const EXIST    = 1 << 1;
-	const LOCK     = 1 << 2;
+        const ANY      = 0;
+        const NO_EXIST = 1;
+        const EXIST    = 1 << 1;
+        const LOCK     = 1 << 2;
+    }
+
+    /// Flags to configurae [`Map::create_map`] operations.
+    pub struct MapCreateFlags: u32 {
```

Comment:
libbpf-sys already has bindings for these flags (https://github.com/libbpf/libbpf-sys/blob/master/src/bindings.rs#L511-L523). I think we can just reuse those instead of redefining them here

Reply:
I think you are right(the reason why I create `MapCreateFlags` is influend by `MapFlags`). In your opinion, I think the `create_map()` should be like(use `u32` type on `flags`):

```rust
    pub fn create_map(
        name: &str,
        map_type: MapType,
        key_size: u32,
        value_size: u32,
        max_entries: u32,
        flags: u32,
    ) -> Result<Map> { ... }
```

And user can call the `create_map()` like this:

```rust
...
    let mut map = Map::create_map(
        "mymap_test_create_map_and_pin",
        MapType::Hash,
        4,
        8,
        8,
        libbpf_sys::BPF_F_NO_PREALLOC
    )
...
```

Reply:
I saw the next comments and `bpf_map_create_opts` will be better.